### ISSUE TITLE
0.32.8 The database has been closed [90098-197] MySQL

### ISSUE THREAD

**Issue by anshulk** [metabase/metabase#10073]
Running 0.32.8 Jar on ubuntu, it worked fine with earlier versions.

{:message "The database has been closed [90098-197]",
 :type org.h2.jdbc.JdbcSQLException,
 :stacktrace
 ("org.h2.message.DbException.getJdbcSQLException(DbException.java:357)"
  "org.h2.message.DbException.get(DbException.java:179)"
  "org.h2.message.DbException.get(DbException.java:155)"
  "org.h2.message.DbException.get(DbException.java:144)"
  "org.h2.engine.Database.checkPowerOff(Database.java:536)"
  "org.h2.command.Command.executeQuery(Command.java:228)"
  "org.h2.jdbc.JdbcPreparedStatement.executeQuery(JdbcPreparedStatement.java:114)"
  "com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeQuery(NewProxyPreparedStatement.java:379)"
  "clojure.java.jdbc$execute_query_with_params.invokeStatic(jdbc.clj:1072)"
  "clojure.java.jdbc$execute_query_with_params.invoke(jdbc.clj:1066)"
  "clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1095)"
  "clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:1075)"
  "clojure.java.jdbc$query.invokeStatic(jdbc.clj:1164)"
  "clojure.java.jdbc$query.invoke(jdbc.clj:1126)"
  "toucan.db$query.invokeStatic(db.clj:274)"
  "toucan.db$query.doInvoke(db.clj:270)"
  "clojure.lang.RestFn.invoke(RestFn.java:410)"
  "--> models.user$permissions_set.invokeStatic(user.clj:280)"
  "models.user$permissions_set.invoke(user.clj:272)"
  "middleware.session$do_with_current_user$fn__58588.invoke(session.clj:175)"
  "models.interface$current_user_permissions_set.invokeStatic(interface.clj:234)"
  "models.interface$current_user_permissions_set.invoke(interface.clj:234)"
  "models.interface$make_perms_check_fn$_has_perms_QMARK___24078.invoke(interface.clj:249)"
  "models.interface$make_perms_check_fn$_has_perms_QMARK___24078.invoke(interface.clj:247)"
  "models.interface$fn__24043$G__24006__24052.invoke(interface.clj:181)"
  "api.collection$fn__44806$collection_detail__44811$fn__44812.invoke(collection.clj:107)"
  "api.collection$fn__44806$collection_detail__44811.invoke(collection.clj:102)"
  "api.collection$root_collection.invokeStatic(collection.clj:137)"
  "api.collection$root_collection.invoke(collection.clj:136)"
  "api.collection$fn__44857.invokeStatic(collection.clj:142)"
  "api.collection$fn__44857.invoke(collection.clj:139)"
  "middleware.auth$enforce_authentication$fn__59926.invoke(auth.clj:14)"
  "routes$fn__61084$fn__61085.doInvoke(routes.clj:56)"
  "middleware.exceptions$catch_uncaught_exceptions$fn__60017.invoke(exceptions.clj:104)"
  "middleware.exceptions$catch_api_exceptions$fn__60014.invoke(exceptions.clj:92)"
  "middleware.log$log_api_call$fn__61458$fn__61459.invoke(log.clj:170)"
  "middleware.log$log_api_call$fn__61458.invoke(log.clj:164)"
  "middleware.security$add_security_headers$fn__59980.invoke(security.clj:105)"
  "middleware.json$wrap_json_body$fn__61163.invoke(json.clj:61)"
  "middleware.json$wrap_streamed_json_response$fn__61181.invoke(json.clj:97)"
  "middleware.session$bind_current_user$fn__58593$fn__58594.invoke(session.clj:193)"
  "middleware.session$do_with_current_user.invokeStatic(session.clj:176)"
  "middleware.session$do_with_current_user.invoke(session.clj:170)"
  "middleware.session$bind_current_user$fn__58593.invoke(session.clj:192)"
  "middleware.session$wrap_current_user_id$fn__58582.invoke(session.clj:161)"
  "middleware.session$wrap_session_id$fn__58567.invoke(session.clj:123)"
  "middleware.auth$wrap_api_key$fn__59934.invoke(auth.clj:27)"
  "middleware.misc$maybe_set_site_url$fn__61523.invoke(misc.clj:58)"
  "middleware.misc$bind_user_locale$fn__61526.invoke(misc.clj:74)"
  "middleware.misc$add_content_type$fn__61511.invoke(misc.clj:30)"
  "middleware.misc$disable_streaming_buffering$fn__61548.invoke(misc.clj:119)"
  "middleware.misc$wrap_gzip$fn__61541.invoke(misc.clj:101)"),
 :sql-exception-chain ["JdbcSQLException:" "Message: The database has been closed [90098-197]" "SQLState: 90098" "Error Code: 90098"]}

**Comment by flamber** [comment]
Hi @anshulk 
Do you see any other warnings or errors before this error?
Where is your H2 database stored - locally on the same disk or externally via NFS/etc?
It's not recommended to use H2 in production - you should migrate to Postgres or MariaDB/MySQL:
https://metabase.com/docs/latest/operations-guide/start.html#migrating-from-using-the-h2-database-to-mysql-or-postgres

**Comment by noamgat** [comment]
I am also getting a similar error - redshift queries randomly fail with a "not connected" error and succeed upon retry. However, this causes pulses to sometimes fail, which is annoying.

**Comment by anshulk** [comment]
@flamber Migrated h2 to Mysql, h2 was on a local file.

Getting this error now - 
```
06-03 07:25:04 ERROR middleware.results-metadata :: Error recording results metadata for query:
 java.sql.BatchUpdateException (conn=11584905) Interrupted awaiting response
 ("org.mariadb.jdbc.MariaDbStatement.executeBatchExceptionEpilogue(MariaDbStatement.java:289)"
 "org.mariadb.jdbc.MariaDbPreparedStatementClient.executeBatch(MariaDbPreparedStatementClient.java:306)"
 "com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeBatch(NewProxyPreparedStatement.java:2437)"
 "clojure.java.jdbc$execute_batch.invokeStatic(jdbc.clj:591)"
 "clojure.java.jdbc$execute_batch.invoke(jdbc.clj:584)"
 "clojure.java.jdbc$db_do_execute_prepared_statement$fn__14358.invoke(jdbc.clj:1039)"
 "clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:799)"
 "clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:769)"
 "clojure.java.jdbc$db_transaction_STAR_.invokeStatic(jdbc.clj:782)"
 "clojure.java.jdbc$db_transaction_STAR_.invoke(jdbc.clj:769)"
 "clojure.java.jdbc$db_do_execute_prepared_statement.invokeStatic(jdbc.clj:1038)"
 "clojure.java.jdbc$db_do_execute_prepared_statement.invoke(jdbc.clj:1024)"
 "clojure.java.jdbc$db_do_prepared.invokeStatic(jdbc.clj:1062)"
 "clojure.java.jdbc$db_do_prepared.invoke(jdbc.clj:1042)"
 "clojure.java.jdbc$execute_BANG_$execute_helper__14426.invoke(jdbc.clj:1446)"
 "clojure.java.jdbc$execute_BANG_.invokeStatic(jdbc.clj:1450)"
 "clojure.java.jdbc$execute_BANG_.invoke(jdbc.clj:1417)"
 "toucan.db$execute_BANG_.invokeStatic(db.clj:411)"
 "toucan.db$execute_BANG_.doInvoke(db.clj:406)"
 "clojure.lang.RestFn.invoke(RestFn.java:410)"
 "toucan.db$update_BANG_.invokeStatic(db.clj:466)"
 "toucan.db$update_BANG_.invoke(db.clj:455)"
 "toucan.db$update_BANG_.invokeStatic(db.clj:475)"
 "toucan.db$update_BANG_.invoke(db.clj:455)"
 "toucan.db$update_BANG_.invokeStatic(db.clj:483)"
 "toucan.db$update_BANG_.doInvoke(db.clj:455)"
 "clojure.lang.RestFn.invoke(RestFn.java:470)"
 "--> query_processor.middleware.results_metadata$record_metadata_BANG_.invokeStatic(results_metadata.clj:23)"
 "query_processor.middleware.results_metadata$record_metadata_BANG_.invoke(results_metadata.clj:21)"
 "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__40249.invoke(results_metadata.clj:97)"
 "query_processor.middleware.format_rows$format_rows$fn__34625.invoke(format_rows.clj:26)"
 "query_processor.middleware.add_dimension_projections$add_remapping$fn__31752.invoke(add_dimension_projections.clj:234)"
 "query_processor.middleware.add_row_count_and_status$add_row_count_and_status$fn__32146.invoke(add_row_count_and_status.clj:16)"
 "query_processor.middleware.driver_specific$process_query_in_context$fn__34386.invoke(driver_specific.clj:12)"
 "query_processor.middleware.resolve_driver$resolve_driver$fn__37215.invoke(resolve_driver.clj:15)"
 "query_processor.middleware.bind_effective_timezone$bind_effective_timezone$fn__33550$fn__33551.invoke(bind_effective_timezone.clj:9)"
 "util.date$call_with_effective_timezone.invokeStatic(date.clj:88)"
 "util.date$call_with_effective_timezone.invoke(date.clj:77)"
 "query_processor.middleware.bind_effective_timezone$bind_effective_timezone$fn__33550.invoke(bind_effective_timezone.clj:8)"
 "query_processor.middleware.store$initialize_store$fn__40274$fn__40275.invoke(store.clj:11)"
 "query_processor.store$do_with_new_store.invokeStatic(store.clj:35)"
 "query_processor.store$do_with_new_store.invoke(store.clj:31)"
 "query_processor.middleware.store$initialize_store$fn__40274.invoke(store.clj:10)"
 "query_processor.middleware.async$async__GT_sync$fn__33146.invoke(async.clj:19)"
 "query_processor.middleware.async_wait$runnable$fn__33277.invoke(async_wait.clj:71)")
```

**Comment by flamber** [comment]
@anshulk 
Okay, then I don't understand why the error log in the first post was referring to H2.
Please describe your entire setup and versions.
And does the error happen constantly or only sometimes?

@noamgat 
Unless you provide the logs, it's difficult to help. I think you might be seeing issue #9989

**Comment by mariummalik22** [comment]
Similar issue at my end as well:

Logs:

message: "The database has been closed [90098-197]"
sql-exception-chain: (4) ["JdbcSQLException:", "Message: The database has been closed [90098-197]", "SQLState: 90098", "Error Code: 90098"]
stacktrace: (48) ["org.h2.message.DbException.getJdbcSQLException(DbException.java:357)", "org.h2.message.DbException.get(DbException.java:179)", "org.h2.message.DbException.get(DbException.java:155)", "org.h2.message.DbException.get(DbException.java:144)", "org.h2.engine.Database.checkPowerOff(Database.java:536)", "org.h2.command.Command.executeQuery(Command.java:228)", "org.h2.jdbc.JdbcPreparedStatement.executeQuery(JdbcPreparedStatement.java:114)", "com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeQuery(NewProxyPreparedStatement.java:379)", "clojure.java.jdbc$execute_query_with_params.invokeStatic(jdbc.clj:1072)", "clojure.java.jdbc$execute_query_with_params.invoke(jdbc.clj:1066)", "clojure.java.jdbc$db_query_with_resultset_STAR_.invokeStatic(jdbc.clj:1095)", "clojure.java.jdbc$db_query_with_resultset_STAR_.invoke(jdbc.clj:1075)", "clojure.java.jdbc$query.invokeStatic(jdbc.clj:1164)", "clojure.java.jdbc$query.invoke(jdbc.clj:1126)", "toucan.db$query.invokeStatic(db.clj:274)", "toucan.db$query.doInvoke(db.clj:270)", "clojure.lang.RestFn.invoke(RestFn.java:410)", "--> models.user$permissions_set.invokeStatic(user.clj:280)", "models.user$permissions_set.invoke(user.clj:272)", "middleware.session$do_with_current_user$fn__58588.invoke(session.clj:175)", "models.interface$current_user_permissions_set.invokeStatic(interface.clj:234)", "models.interface$current_user_permissions_set.invoke(interface.clj:234)", "models.interface$make_perms_check_fn$_has_perms_QMARK___24078.invoke(interface.clj:249)", "models.interface$make_perms_check_fn$_has_perms_QMARK___24078.invoke(interface.clj:247)", "models.interface$fn__24013$G__24004__24022.invoke(interface.clj:181)", "api.collection$fn__44863.invokeStatic(collection.clj:162)", "api.collection$fn__44863.invoke(collection.clj:144)", "middleware.auth$enforce_authentication$fn__59926.invoke(auth.clj:14)", "routes$fn__61084$fn__61085.doInvoke(routes.clj:56)", "middleware.exceptions$catch_uncaught_exceptions$fn__60017.invoke(exceptions.clj:104)", "middleware.exceptions$catch_api_exceptions$fn__60014.invoke(exceptions.clj:92)", "middleware.log$log_api_call$fn__61458$fn__61459.invoke(log.clj:170)", "middleware.log$log_api_call$fn__61458.invoke(log.clj:164)", "middleware.security$add_security_headers$fn__59980.invoke(security.clj:105)", "middleware.json$wrap_json_body$fn__61163.invoke(json.clj:61)", "middleware.json$wrap_streamed_json_response$fn__61181.invoke(json.clj:97)", "middleware.session$bind_current_user$fn__58593$fn__58594.invoke(session.clj:193)", "middleware.session$do_with_current_user.invokeStatic(session.clj:176)", "middleware.session$do_with_current_user.invoke(session.clj:170)", "middleware.session$bind_current_user$fn__58593.invoke(session.clj:192)", "middleware.session$wrap_current_user_id$fn__58582.invoke(session.clj:161)", "middleware.session$wrap_session_id$fn__58567.invoke(session.clj:123)", "middleware.auth$wrap_api_key$fn__59934.invoke(auth.clj:27)", "middleware.misc$maybe_set_site_url$fn__61523.invoke(misc.clj:58)", "middleware.misc$bind_user_locale$fn__61526.invoke(misc.clj:74)", "middleware.misc$add_content_type$fn__61511.invoke(misc.clj:30)", "middleware.misc$disable_streaming_buffering$fn__61548.invoke(misc.clj:119)", "middleware.misc$wrap_gzip$fn__61541.invoke(misc.clj:101)"]
type: "class org.h2.jdbc.JdbcSQLException"
__proto__: Object
isCancelled: false
status: 500
__proto__: Object

**Comment by mariummalik22** [comment]
@anshulk Were you able to solve the issue?

**Comment by flamber** [comment]
@mariummalik22 
Which version of Metabase? The latest release is 0.32.10
From looking at the error message, it looks like you're using the default H2 as internal metadata database. Is that correct? If yes, then you should migrate away:
https://metabase.com/docs/latest/operations-guide/start.html#migrating-from-using-the-h2-database-to-mysql-or-postgres

**Comment by jane-booky** [comment]
Hi @flamber , I'm sorry could you give a quick explanation for the need to migrate from H2?
I am also experiencing the error above frequently, I wanted to have it retain at H2 due to certain setup limitations for now, but if it greatly affects performance / if it will solve the issue we are having right now, then we will migrate.

**Comment by flamber** [comment]
Hi @jane-booky

Besides the above error, when H2 is being put under load? Well, if you search the forums you'll find many posts, where the H2 ended up corrupted because of crash/power issues.

Since Metabase cannot do regular H2 [cleanup](http://www.h2database.com/html/features.html#compacting) while it's running, then you would have to schedule Metabase service restarts to keep it somewhat optimized (limited automatic since Metabase 0.32, before that or more extensive cleanup, you would have to run commands manually).

And H2 only allows write access to a single process, so that means Metabase is limited to run in a single instance, instead of running multiple Metabase instances (horizontal scalable) connected to the same Postgres/MariaDB/MySQL metadata.

But without knowing more details like your usage (amount of active users, database, size of metadata, etc) and resources (CPU usage, RAM usage, etc) and environment (Metabase version, OS, Java, etc), then it's difficult to know why you might be encountering this issue.

**Comment by camsaul** [comment]
Fixed by #11832, which correctly holds JDBC result sets open until all rows have been parsed
